db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java

```diff
@@ -322,42 +365,77 @@ public <T> void schedule(SchedulableInstance<T> schedulableInstance) {
     }
 
     @Override
-    public void reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTime) {
-      reschedule(taskInstanceId, newExecutionTime, null);
+    public <T> boolean schedule(
+        TaskInstance<T> taskInstance, Instant executionTime, WhenExists whenExists) {
+      Optional<ScheduledExecution<Object>> executionObject =
+          this.getScheduledExecution(taskInstance);
+      if (executionObject.isPresent()) {
+        switch (whenExists) {
+          case RESCHEDULE:
+            LOG.info("Task instance already exists. Rescheduling task instance: {}", taskInstance);
+            return this.reschedule(taskInstance, executionTime, taskInstance.getData());
+          case DO_NOTHING:
+            LOG.info(
+                "Task instance already exists. No further action taken for task instance: {}",
+                taskInstance);
+            return true;
+          default:
+            throw new IllegalArgumentException("Unknown WhenExists: " + whenExists);
+        }
+      }
+
+      LOG.info("Scheduling new task instance: {}", taskInstance);
+      return this.scheduleIfNotExists(taskInstance, executionTime);
+    }
+
+    @Override
+    public <T> boolean schedule(SchedulableInstance<T> schedulableInstance, WhenExists whenExists) {
+      return schedule(
+          schedulableInstance.getTaskInstance(),
+          schedulableInstance.getNextExecutionTime(clock.now()),
+          whenExists);
     }
 
     @Override
-    public <T> void reschedule(SchedulableInstance<T> schedulableInstance) {
-      reschedule(
+    public boolean reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTime) {
+      return reschedule(taskInstanceId, newExecutionTime, null);
+    }
+
+    @Override
+    public <T> boolean reschedule(SchedulableInstance<T> schedulableInstance) {
+      return reschedule(
           schedulableInstance,
           schedulableInstance.getNextExecutionTime(clock.now()),
           schedulableInstance.getTaskInstance().getData());
     }
 
     @Override
-    public <T> void reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTime, T newData) {
+    public <T> boolean reschedule(
+        TaskInstanceId taskInstanceId, Instant newExecutionTime, T newData) {
       String taskName = taskInstanceId.getTaskName();
       String instanceId = taskInstanceId.getId();
-      Optional<Execution> execution = taskRepository.getExecution(taskName, instanceId);
-      if (execution.isPresent()) {
-        if (execution.get().isPicked()) {
-          throw new TaskInstanceCurrentlyExecutingException(taskName, instanceId);
-        }
+      Execution execution =
+          taskRepository
+              .getExecution(taskName, instanceId)
+              .orElseThrow(() -> new TaskInstanceNotFoundException(taskName, instanceId));
```

Comment:
With the use of `orElseThrow` we can get rid of the big brackets which goes around 10+ lines of code, which should improve the code readability.